engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java

```diff
@@ -0,0 +1,836 @@
+        .sorted(
+            (entry1, entry2) -> {
+              var fqn1 = entry1.getKey();
+              var fqn2 = entry2.getKey();
+              if (fqn1.equals(defCompFQN)) {
+                return -1;
+              } else if (fqn2.equals(defCompFQN)) {
+                return 1;
+              } else {
+                return fqn1.compareTo(fqn2);
+              }
+            })
+        .map(Entry::getValue)
+        .collect(Collectors.toList());
+  }
+
+  private Object attachWarnings(Object vector, Set<String> warnings) {
+    var warnArray =
+        warnings.stream()
+            .map(Text::create)
+            .map(text -> Warning.create(EnsoContext.get(this), text, this))
+            .toArray(Warning[]::new);
+    return WithWarnings.appendTo(vector, new ArrayRope<>(warnArray));
+  }
+
+  private Object attachDifferentComparatorsWarning(Object vector, List<Group> groups) {
+    var diffCompsMsg =
+        groups.stream()
+            .map(Group::comparator)
+            .map(comparator -> comparator.getQualifiedName().toString())
+            .collect(Collectors.joining(", "));
+    var text = Text.create("Different comparators: [" + diffCompsMsg + "]");
+    logger.finer(
+        () -> String.format("Will attach different comparators warning '%s'", diffCompsMsg));
+    var warn = Warning.create(EnsoContext.get(this), text, this);
+    return WithWarnings.appendTo(vector, new ArrayRope<>(warn));
+  }
+
+  private String getDefaultComparatorQualifiedName() {
+    return EnsoContext.get(this)
+        .getBuiltins()
+        .defaultComparator()
+        .getType()
+        .getQualifiedName()
+        .toString();
+  }
+
+  /** A group is "primitive" iff its comparator is the default comparator. */
+  private boolean isPrimitiveGroup(Group group) {
+    return group
+        .comparator
+        .getQualifiedName()
+        .toString()
+        .equals(getDefaultComparatorQualifiedName());
+  }
+
+  private Object incomparableValuesError(Object left, Object right) {
+    return EnsoContext.get(this).getBuiltins().error().makeIncomparableValues(left, right);
+  }
+
+  /**
+   * Helper slow-path method to conveniently gather elements from interop arrays into a java list
+   */
+  @SuppressWarnings("unchecked")
+  private <T> List<T> readInteropArray(
+      InteropLibrary interop, WarningsLibrary warningsLib, Object vector) {
+    try {
+      int size = (int) interop.getArraySize(vector);
+      List<T> res = new ArrayList<>(size);
+      for (int i = 0; i < size; i++) {
+        Object elem = interop.readArrayElement(vector, i);
+        if (warningsLib.hasWarnings(elem)) {
+          elem = warningsLib.removeWarnings(elem);
+        }
+        res.add((T) elem);
+      }
+      return res;
+    } catch (UnsupportedMessageException | InvalidArrayIndexException | ClassCastException e) {
+      throw new IllegalStateException("Should not be reachable", e);
+    }
+  }
+
+  private int getBuiltinTypeCost(Object builtinType) {
```

Comment:
What is the meaning of _cost_ in this context? Please document.

Reply:
Is `Cost` the right name here?

Reply:
Two values are assigned an integer *cost*, and the one that has lower *cost* is lower. I guess that `weight` would also be acceptable name. Any better name suggestions? If you look into the call site of this method, it is pretty clear what it does and how it should behave.

Reply:
Yeah. Weight sounds better IMO.
I guess you are right, maybe a note that this weight determines the relative ordering of each of primitive types?